src/rust/engine/src/rule_graph.rs

```diff
@@ -1151,18 +1111,15 @@ impl RuleEdges {
 ///
 /// Select Entries that can provide the given product type with the given parameters.
 ///
-fn rhs(tasks: &Tasks, params: &ParamTypes, product_type: &TypeConstraint) -> Vec<Entry> {
+fn rhs(tasks: &Tasks, params: &ParamTypes, product_type: TypeId) -> Vec<Entry> {
   if let Some(&(ref key, _)) = tasks.gen_singleton(product_type) {
-    return vec![Entry::Singleton(*key, *product_type)];
+    return vec![Entry::Singleton(*key, product_type)];
   }
 
   let mut entries = Vec::new();
-  if let Some(type_id) = params
-    .iter()
-    .find(|&&type_id| externs::satisfied_by_type(product_type, type_id))
-  {
+  if let Some(type_id) = params.get(&product_type) {
     // TODO: We only match the first param type here that satisfies the constraint although it's
```

Comment:
I think the whole TODO can now be removed! Yay.

Reply:
Done! Added some brief comments too.